addons/binding/org.openhab.binding.plugwise/src/main/java/org/openhab/binding/plugwise/internal/PlugwiseCommunicationHandler.java

```diff
@@ -0,0 +1,591 @@
+            }
+        }
+        return false;
+    }
+
+    private void notifyListeners(Message message) {
+        for (FilteredMessageListener filteredListener : filteredListeners) {
+            if (filteredListener.matches(message)) {
+                try {
+                    filteredListener.listener.handleReponseMessage(message);
+                } catch (Exception e) {
+                    logger.warn("Listener failed to handle message: {}", message, e);
+                }
+            }
+        }
+    }
+
+    /**
+     * Parse a buffer into a Message and put it in the appropriate queue for further processing
+     *
+     * @param readBuffer - the string to parse
+     */
+    private void parseAndQueue(ByteBuffer readBuffer) {
+        if (readBuffer != null) {
+
+            String response = new String(readBuffer.array(), 0, readBuffer.limit());
+            response = StringUtils.chomp(response);
+
+            Matcher matcher = RESPONSE_PATTERN.matcher(response);
+
+            if (matcher.matches()) {
+
+                String protocolHeader = matcher.group(1);
+                String messageTypeHex = matcher.group(2);
+                String sequence = matcher.group(3);
+                String payload = matcher.group(4);
+                String crc = matcher.group(5);
+
+                if (protocolHeader.equals(PROTOCOL_HEADER)) {
+                    String calculatedCRC = Message.getCRC(messageTypeHex + sequence + payload);
+                    if (calculatedCRC.equals(crc)) {
+                        MessageType messageType = MessageType.forValue(Integer.parseInt(messageTypeHex, 16));
+                        int sequenceNumber = Integer.parseInt(sequence, 16);
+
+                        if (messageType == null) {
+                            logger.debug("Received unrecognized message: messageTypeHex=0x{}, sequence={}, payload={}",
+                                    messageTypeHex, sequenceNumber, payload);
+                            return;
+                        }
+
+                        logger.debug("Received message: messageType={}, sequenceNumber={}, payload={}", messageType,
+                                sequenceNumber, payload);
+
+                        try {
+                            Message message = messageFactory.createMessage(messageType, sequenceNumber, payload);
+
+                            if (message instanceof AcknowledgementMessage
+                                    && !((AcknowledgementMessage) message).isExtended()) {
+                                logger.debug("Adding to acknowledgedQueue: {}", message);
+                                acknowledgedQueue.put((AcknowledgementMessage) message);
+                            } else {
+                                logger.debug("Adding to receivedQueue: {}", message);
+                                receivedQueue.put(message);
+                            }
+                        } catch (IllegalArgumentException e) {
```

Comment:
The MessageFactory throws it when it fails to create unknown messages. 

This code now lives in the PlugwiseMessageProcessor.